test/include/seqan3/test/tmp_filename_test.cpp

```diff
@@ -0,0 +1,89 @@
+// ==========================================================================
+//                 SeqAn - The Library for Sequence Analysis
+// ==========================================================================
+//
+// Copyright (c) 2006-2017, Knut Reinert, FU Berlin
+// Copyright (c) 2016-2017, Knut Reinert & MPI Molekulare Genetik
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ==========================================================================
+
+#include <gtest/gtest.h>
+#include <fstream>
+#include "tmp_filename.hpp"
+
+using namespace seqan3::test;
+
+// aggregate initialization
+TEST(tmp_filename_aggr, aggr)
+{
+    tmp_file_name t1{"aggr_test"};
+    tmp_file_name t2("aggr_test");
+    EXPECT_NE(t1.get_path(), t2.get_path());
```

Comment:
Do we expect nothing more of a constructed `tmp_file_name`, something like the base of `get_path()` must be equal to tmp directory?

Reply:
Mhm, we could test if the temporary directory exists, but ``fs::temp_directory_path`` will throw if it doesn't exist.
In general, the file paths look like
``/tmp/seqan_test_XXXXXXXX/filename``
Where the X's are random numbers that are supposed to be unique.

And is also doesn't have to be ``/tmp`` since you can define env vars to change the tmp dir.
See http://en.cppreference.com/w/cpp/filesystem/temp_directory_path

Reply:
If the construction might throw I would add
```cpp
EXPECT_NO_THROW(tmp_file_name t1{"aggr_test"}); 
```
I wasn't sure if the directory and the file are directly created or rather on demand. If they are created on construction I would definitely check for existence afterwards.

Thanks for the short explanation. I meant to maybe check for `base_path == fs::temp_directory_path()`.

Do you know how the random numbers are generated? I'm curious if those are always unique and thread safe. Care to try out something like this?:

```cpp
// TEST_START
std::vector<tmp_file_name> filenames;

for (unsigned i = 0 ;i < 10000; ++i)
    filenames.push_back(tmp_file_name("test"));

std::sort(filenames.begin(), filenames.end(), [] (tmp_file_name const & lhs, tmp_file_name const & rhs) {return lhs.get_path() < rhs.get_path; });
auto last = std::unique(filenames.begin(), filenames.end(), [] (tmp_file_name const & lhs, tmp_file_name const & rhs) {return lhs.get_path() == rhs.get_path; } );

EXPECT_EQ(last, filenames.end()); // no unique elements
```
(I don't know if the lambdas work, maybe you need a functor )

Reply:
I added two test to the construction to check whether the directory is created on construction and the base path is the temporary dir.

I tested the other thing, and it works as promised by the specification of mkdtemp.
It would fail if we created more than 10.000.000 temporary files.

Regarding thread safety, depending where you check, mkdtemp is either thread safe or it is not documented.

Reply:
mkdtemp is guaranteed to be thread-safe

>Function: char * mkdtemp (char *template)
Preliminary: | MT-Safe | AS-Safe | AC-Safe | See POSIX Safety Concepts.

Reply:
Should we still include this as a test case? (1) test 1000 file names (2) test threading. 
Even if this is maintained by mkdtemp it would be beneficial to notice failures right away.